File: FreeLine-Android/app/src/main/java/com/freeline/app/messaging/MessageModels.kt
package com.freeline.app.messaging

import org.json.JSONObject

data class MessageAllowance(
    val dailyCap: Int,
    val dailyRemaining: Int,
    val dailyUsed: Int,
    val monthlyCap: Int,
    val monthlyRemaining: Int,
    val monthlyUsed: Int,
)

data class ConversationSummary(
    val createdAt: String,
    val id: String,
    val isOptedOut: Boolean,
    val lastMessageAt: String?,
    val lastMessagePreview: String?,
    val lastMessageStatus: String?,
    val lastSpamConfidence: Double?,
    val lastSpamReason: String?,
    val participantNumber: String,
    val phoneNumberId: String,
    val unreadCount: Int,
    val updatedAt: String,
    val userId: String,
) {
    val displayNumber: String
        get() = participantNumber.formatUsPhoneNumber()

    val isLastMessageSpam: Boolean
        get() = (lastSpamConfidence ?: 0.0) >= 0.6
}

data class ChatMessage(
    val body: String,
    val conversationId: String,
    val createdAt: String,
    val direction: String,
    val id: String,
    val providerMessageId: String?,
    val spamConfidence: Double?,
    val spamReason: String?,
    val status: String,
    val updatedAt: String,
) {
    val isOutgoing: Boolean
        get() = direction == "outbound"

    val isLikelySpam: Boolean
        get() = (spamConfidence ?: 0.0) >= 0.6

    val spamBadgeText: String?
        get() {
            val conf = spamConfidence ?: return null
            if (conf < 0.5) return null
            return "Spam ${(conf * 100).toInt()}%"
        }
}

data class ConversationListPayload(
    val allowance: MessageAllowance,
    val conversations: List<ConversationSummary>,
    val limit: Int,
    val offset: Int,
)

data class ConversationThreadPayload(
    val allowance: MessageAllowance,
    val conversation: ConversationSummary,
    val limit: Int,
    val messages: List<ChatMessage>,
    val offset: Int,
)

data class MessageSendPayload(
    val allowance: MessageAllowance,
    val conversation: ConversationSummary,
    val message: ChatMessage,
)

enum class MessageRealtimeEventType(val wireName: String) {
    Ready("realtime:ready"),
    MessageInbound("message:inbound"),
    MessageStatus("message:status");

    companion object {
        fun fromWireName(value: String): MessageRealtimeEventType? = entries.firstOrNull { it.wireName == value }
    }
}

data class MessageRealtimeEvent(
    val conversation: ConversationSummary?,
    val message: ChatMessage?,
    val type: MessageRealtimeEventType,
)

fun normalizeUsPhoneNumber(rawValue: String): String? {
    val digits = rawValue.filter(Char::isDigit)

    return when {
        rawValue.startsWith("+") && digits.length == 11 && digits.startsWith("1") -> "+$digits"
        digits.length == 10 -> "+1$digits"
        digits.length == 11 && digits.startsWith("1") -> "+$digits"
        else -> null
    }
}

fun String.formatUsPhoneNumber(): String {
    val digits = filter(Char::isDigit)

    if (digits.length != 11 || !digits.startsWith("1")) {
        return this
    }

    val areaCode = digits.substring(1, 4)
    val prefix = digits.substring(4, 7)
    val lineNumber = digits.substring(7, 11)
    return "($areaCode) $prefix-$lineNumber"
}

internal fun JSONObject.toConversation(): ConversationSummary =
    ConversationSummary(
        createdAt = getString("createdAt"),
        id = getString("id"),
        isOptedOut = getBoolean("isOptedOut"),
        lastMessageAt = optString("lastMessageAt").ifBlank { null },
        lastMessagePreview = optString("lastMessagePreview").ifBlank { null },
        lastMessageStatus = optString("lastMessageStatus").ifBlank { null },
        lastSpamConfidence = if (has("lastSpamConfidence") && !isNull("lastSpamConfidence")) optDouble("lastSpamConfidence") else null,
        lastSpamReason = if (has("lastSpamReason") && !isNull("lastSpamReason")) optString("lastSpamReason").ifBlank { null } else null,
        participantNumber = getString("participantNumber"),
        phoneNumberId = getString("phoneNumberId"),
        unreadCount = getInt("unreadCount"),
        updatedAt = getString("updatedAt"),
        userId = getString("userId"),
    )

internal fun JSONObject.toMessage(): ChatMessage =
    ChatMessage(
        body = getString("body"),
        conversationId = getString("conversationId"),
        createdAt = getString("createdAt"),
        direction = getString("direction"),
        id = getString("id"),
        providerMessageId = optString("providerMessageId").ifBlank { null },
        spamConfidence = if (has("spamConfidence") && !isNull("spamConfidence")) optDouble("spamConfidence") else null,
        spamReason = if (has("spamReason") && !isNull("spamReason")) optString("spamReason").ifBlank { null } else null,
        status = getString("status"),
        updatedAt = getString("updatedAt"),
    )

File: FreeLine-iOS/Sources/Messages/MessageModels.swift
import Foundation

struct MessageAllowance: Codable, Equatable {
    let dailyCap: Int
    let dailyRemaining: Int
    let dailyUsed: Int
    let monthlyCap: Int
    let monthlyRemaining: Int
    let monthlyUsed: Int
}

struct ConversationSummary: Codable, Equatable, Hashable, Identifiable {
    let createdAt: String
    let id: String
    let isOptedOut: Bool
    let lastMessageAt: String?
    let lastMessagePreview: String?
    let lastMessageStatus: String?
    let lastSpamConfidence: Double?
    let lastSpamReason: String?
    let participantNumber: String
    let phoneNumberId: String
    let unreadCount: Int
    let updatedAt: String
    let userId: String

    var displayNumber: String {
        participantNumber.formattedUSPhoneNumber
    }

    var isLastMessageSpam: Bool {
        (lastSpamConfidence ?? 0) >= 0.6
    }
}

struct ChatMessage: Codable, Equatable, Hashable, Identifiable {
    let body: String
    let conversationId: String
    let createdAt: String
    let direction: String
    let id: String
    let providerMessageId: String?
    let spamConfidence: Double?
    let spamReason: String?
    let status: String
    let updatedAt: String

    var isOutgoing: Bool {
        direction == "outbound"
    }

    var isLikelySpam: Bool {
        (spamConfidence ?? 0) >= 0.6
    }

    var spamBadgeText: String? {
        guard let confidence = spamConfidence, confidence >= 0.5 else { return nil }
        let pct = Int(confidence * 100)
        return "Spam \(pct)%"
    }
}

struct ConversationListPayload: Codable, Equatable {
    let allowance: MessageAllowance
    let conversations: [ConversationSummary]
    let limit: Int
    let offset: Int
}

struct ConversationThreadPayload: Codable, Equatable {
    let allowance: MessageAllowance
    let conversation: ConversationSummary
    let limit: Int
    let messages: [ChatMessage]
    let offset: Int
}

struct MessageSendPayload: Codable, Equatable {
    let allowance: MessageAllowance
    let conversation: ConversationSummary
    let message: ChatMessage
}

struct ConversationReadPayload: Codable, Equatable {
    let conversation: ConversationSummary
}

enum MessageRealtimeEventType: String, Codable, Equatable {
    case ready = "realtime:ready"
    case messageInbound = "message:inbound"
    case messageStatus = "message:status"
}

struct MessageRealtimeEvent: Codable, Equatable {
    let conversation: ConversationSummary?
    let message: ChatMessage?
    let type: MessageRealtimeEventType
}

struct BlockRecord: Codable, Equatable {
    let blockedNumber: String
    let createdAt: String
    let id: String
    let userId: String
}

struct BlockPayload: Codable, Equatable {
    let block: BlockRecord
}

struct ReportRecord: Codable, Equatable {
    let createdAt: String
    let id: String
    let reason: String
    let reportedNumber: String
    let userId: String
}

struct ReportPayload: Codable, Equatable {
    let report: ReportRecord
}

extension String {
    var formattedUSPhoneNumber: String {
        let digits = filter(\.isNumber)

        guard digits.count == 11, digits.first == "1" else {
            return self
        }

        let characters = Array(digits)
        let areaCode = String(characters[1...3])
        let prefix = String(characters[4...6])
        let lineNumber = String(characters[7...10])
        return "(\(areaCode)) \(prefix)-\(lineNumber)"
    }
}
